### PR TITLE
Interactive mode for mason new

### LINKED ISSUE #14952: Add `--interactive` option for mason new and mason init
In #14820 and #14786, a need for an `--interactive` tag has been mentioned for `mason init` and `mason new`. Using the commands, `mason init --interactive` or `mason new --interactive` will start a CLI questionnaire , in which the user will be prompted to input some values for `projectName` , `version`, `author` etc. for their library project. If an user leaves the fields empty, then default values will be taken.

### PR DESCRIPTION
**Pull request by ankingcodes**
fixes #14952 

This PR adds an interactive mode to `mason new` when given no package argument.

### CODE REVIEW

File: tools/mason/MasonNew.chpl [hunk 1]
```diff
@@ -119,14 +119,16 @@ proc beginInteractiveSession() throws {
...
       }
       if gotCorrectPackageVersion == false {
         write("Package version (0.1.0): ");
         IO.stdout.flush();
-        readf("%s", version);
+        IO.stdin.readline(version);
+        version = version.strip();
         if version == "" then version = "0.1.0";
         checkVersion(version); 
         gotCorrectPackageVersion = true;
```
Comment (ankingcodes) [inline tools/mason/MasonNew.chpl]: If an user is using `mason new --interactive` for creating their project, I assumed that the version should always be `0.1.0`. However, I haven't added a check to force the user to use `0.1.0`. Should there be a check ?

File: tools/mason/MasonNew.chpl [hunk 2]
```diff
@@ -135,12 +137,13 @@ proc beginInteractiveSession() throws {
...
         write("Chapel version (" + currChapelVersion + "): ");
         IO.stdout.flush();
-        readf("%s", chapelVersion);
-        if chapelVersion == currChapelVersion then gotCorrectChapelVersion = true;
+        IO.stdin.readline(chapelVersion);
+        chapelVersion = chapelVersion.strip();
         if chapelVersion == "" then chapelVersion = currChapelVersion;
         if chapelVersion != currChapelVersion then 
           throw new owned MasonError("This version of Chapel is deprecated. Please use" +
               " latest version " + currChapelVersion);
```
Comment (ankingcodes) [inline tools/mason/MasonNew.chpl]: I've added a check here to make sure that an user is using the latest version of Chapel. Should there be an option for an user to use any other previous version here ? Also, could this error message be improved ?

**Comment by ankingcodes** [comment]
@ben-albrecht Please add the GSOC labels to this PR and the corresponding issue. Also, take a look at the comments above.

File: tools/mason/MasonNew.chpl (hunk 1 above)
Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl]: I don't think so. I think `0.1.0` makes sense as a default, but may a user has a reason to start with a different version.

File: tools/mason/MasonNew.chpl (hunk 2 above)
Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl]: I think so. We should probably assume the user knows what they are doing if they override the default. 

We should probably print a warning if the user selects a version range that is not compatible with their current version, e.g. `0.1.3` or `1.19..1.20`

Comment (ankingcodes) [reply tools/mason/MasonNew.chpl]: @ben-albrecht Are users allowed to use a version range for `chplVersion` ? 
Also, how would compatibility of versions be checked ? Can you elaborate this ?

Comment (ankingcodes) [reply tools/mason/MasonNew.chpl]: We can replace the `MasonError` message here with a `Note` which would advice users to use latest version of Chapel

Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl]: > Are users allowed to use a version range for chplVersion ?
> Also, how would compatibility of versions be checked ? Can you elaborate this ?

Yes, it uses range syntax, where no range implies a "this version or above", or `x...`. Examples: `1.19.0` == `1.19.0..`, `1.20..1.21`, or `1.22.0..1.22.0` (only works with 1.22)

Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl]: > Also, how would compatibility of versions be checked ? 

I think we just need to check that the current chapel version is within the specified range.

Comment (ankingcodes) [reply tools/mason/MasonNew.chpl]: @ben-albrecht I found that this check is already done by `mason update`. I feel this feature is not very necessary for interactive, since, we are already providing default chapel version hint while taking input. If an user makes a mistake and uses an incompatible version, I believe it should be `mason update` and `mason build` which should display them the error message rather than interactive.

Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl]: Would it be difficult to call the same validation function already used in `mason update` in the interactive session?

I agree that it's somewhat redundant, but it could also be a nice way to quickly catch and correct a user error, rather than letting them complete the interactive setup and learn of their mistake afterwards.

For example:

```sh
> mason new --interactive
This is an interactive session to walk you through creating a library project using Mason.
The following queries covers the common items required to create the project. 
It also tries to guess sensible defaults.

Press <somebutton> to quit. 


name (default displayed here) : BensPkg
version (default displayed) :  0.1.0
chplVersion (default displayed) : 1.20.0..1.22.0
Warning: Your current Chapel version (1.23.0) is not compatible with this chplVersion.
Are you sure you want to use: 1.20.0..1.22.0 ? [Y/n]

...
```

It's a bit of a UI micro-optimization, so I'm fine either way.

**Comment by ankingcodes** [comment]
@ben-albrecht This PR is ready for a review now 👍 
Will update documentation post-review.

File: tools/mason/MasonNew.chpl [hunk 3]
```diff
@@ -91,13 +103,112 @@ proc masonNew(args) throws {
   }
 }
 
+/*
+  Starts an interactive session to create a
+  new library project.
+*/
+proc beginInteractiveSession(isInit: bool, defaultPackageName: string) throws {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:110]: Another way we could do this is:

```chpl
proc beginInteractiveSession(defaultPackageName='') throws {
  ...
}
```

If a package name is passed to the function, we can assume we are using `mason init`. If not, we can assume we are using `mason new`.

File: tools/mason/MasonNew.chpl [hunk 4]
```diff
@@ -91,13 +103,112 @@ proc masonNew(args) throws {
...
+  writeln("This is an interactive session to walk you through creating a library ",
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:111]: Note that `"""multi-line strings"""` are available in Chapel, which may make this a little cleaner

File: tools/mason/MasonNew.chpl [hunk 5]
```diff
@@ -91,13 +103,112 @@ proc masonNew(args) throws {
...
+  var gotCorrectPackageName = false;
+  var gotCorrectPackageVersion = false;
+  var gotCorrectChapelVersion = false;
+  while(1){
+    try {
+      if gotCorrectPackageName == false {
+        write("Package name ");
+        if isInit == true then write("(" + defaultPackageName + ") ");
+        write(": ");
+        IO.stdout.flush();
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:128]: No change needed, but note:

When using a package (`use IO;`), we do not require qualified access to module identifiers, so you could also just write `stdout.flush();`.

When importing a package (`import IO;`), you must qualify module identifiers like you have done here. This is new as of Chapel 1.21.0.

I personally prefer qualified accesses most of the time, so I like what you've done here - just making sure you are aware.

File: tools/mason/MasonNew.chpl [hunk 6]
```diff
@@ -91,13 +103,112 @@ proc masonNew(args) throws {
...
+        IO.stdin.readline(packageName);
+        packageName = packageName.strip();
+        if isInit == true && packageName == '' then
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:131]:
```suggestion
        if isInit && packageName == '' then
```

File: tools/mason/MasonNew.chpl [hunk 7]
```diff
@@ -91,13 +103,112 @@ proc masonNew(args) throws {
...
+          packageName = defaultPackageName; 
+        if validatePackageName(packageName) == true then
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:133]:
```suggestion
        if validatePackageName(packageName) then
```

File: tools/mason/MasonNew.chpl [hunk 8]
```diff
@@ -91,13 +103,112 @@ proc masonNew(args) throws {
...
+          gotCorrectPackageName = true;
+      }
+      if gotCorrectPackageVersion == false {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:136]:
```suggestion
      if !gotCorrectPackageVersion {
```

Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl:136]: This applies elsewhere, but I'll stop commenting on them..

File: tools/mason/MasonNew.chpl [hunk 9]
```diff
@@ -91,13 +103,112 @@ proc masonNew(args) throws {
...
+      writeln(e.message());
+      continue;
+    }
+  }
+  return (packageName, version, chapelVersion);
+}
+
+/* Previews the Mason.toml file that is going to be created */
+proc previewMasonFile(packageName, version, chapelVersion) {
+  const baseToml = '[brick]\n' +
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:180]: A multi-line string + `.format()` could be cleaner here , e.g.

```chpl
const baseToml = '''[brick]
name = "%s"
...
[dependencies]
'''.format(packageName, version, chapelVersion);
```

Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl:180]: If you make this function just return the string, could you re-use it in `makeBasicToml`?

File: tools/mason/MasonUpdate.chpl [hunk 10]
```diff
@@ -196,7 +196,23 @@ proc parseChplVersion(brick: borrowed Toml?): (VersionInfo, VersionInfo) {
...
+    const name = brick!["name"]!.s + "-" + brick!["version"]!.s;
+    stderr.writeln("Invalid chplVersion in package '", name, "': ", chplVersion);
+    stderr.writeln("Details: ", e.message());
+    exit(1);
+  }
+
+  return (low, hi);
+}
+
+proc checkChplVersion(chplVersion, low, hi) throws {
```
Comment (ben-albrecht) [inline tools/mason/MasonUpdate.chpl:212]: How about we use `low, high` and `lo, hi` for arg/variable names, respectively? This lets us stick to typical Chapel style of camelCase vars

File: tools/mason/.gitignore [hunk 11]
```diff
@@ -1,2 +1,3 @@
 mason
 mason_real
+
```
[the remark this answers is not in the document]
Comment (ankingcodes) [reply tools/mason/.gitignore:3]: I had deleted the file by mistake. I believe a newline should be there anyways.

File: tools/mason/MasonNew.chpl (hunk 4 above)
Comment (ankingcodes) [reply tools/mason/MasonNew.chpl:111]: @ben-albrecht I did try to use that, but I had to remove whitespaces before the second and next lines which would break our indentation. Is there a way to ignore whitespaces  ?

Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl:111]: I'm working on it...  https://github.com/chapel-lang/chapel/pull/15664

See https://github.com/chapel-lang/chapel/issues/8281 for background if curious.

I would be OK breaking indentation for now, and we can come back and re-indent whenever `dedent` is available.

**Comment by ankingcodes** [comment]
@ben-albrecht made changes requested. Let me know for changes. 👍

File: tools/mason/MasonNew.chpl [hunk 12]
```diff
@@ -91,13 +103,120 @@ proc masonNew(args) throws {
...
+      if gotCorrectPackageName &&
+         gotCorrectPackageVersion &&
+         gotCorrectChapelVersion {
+          previewMasonFile(packageName, version, chapelVersion);
+          writeln();
+          write("Is this okay ? (Y/N): ");
+          IO.stdout.flush();
+          var option: string;
+          IO.stdin.readline(option);
+          option = option.strip();
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:171]: Should this be case insensitive?

File: tools/mason/MasonNew.chpl [hunk 13]
```diff
@@ -152,24 +273,39 @@ proc addGitIgnore(dirName: string) {
...
 
-proc makeBasicToml(dirName: string, path: string) {
-  const baseToml = '[brick]\n' +
-                     'name = "' + dirName + '"\n' +
-                     'version = "0.1.0"\n' +
-                     'chplVersion = "' + getChapelVersionStr() + '"\n' +
-                     '\n' +
-                     '[dependencies]' +
-                     '\n';
+proc getBaseTomlString(packageName: string, version: string, chapelVersion: string) {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:276]: Great refactor!

File: tools/mason/MasonNew.chpl (hunk 12 above)
Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl:171]: to support `y/n` and `Y/N`.

File: doc/rst/tools/mason/mason.rst [hunk 14]
```diff
@@ -44,6 +44,9 @@ This creates a git repository by default, unless ``--no-vcs`` is included.
 Mason packages can also be initialized using the ``mason init [options] [directory path]`` or 
 ``mason init [options]`` command outside or inside the project directory respectively. 
 
+There is also an interactive way to create Mason packages using ``mason new --interactive`` or
+``mason init --interactive``. The interactive sessions walks an user through creating a package in Mason.
```
[the remark this answers is not in the document]
Comment (ankingcodes) [reply doc/rst/tools/mason/mason.rst:48]: I think "session" is better since there is only one session. 😄

[the remark this answers is not in the document]
Comment (ben-albrecht) [reply doc/rst/tools/mason/mason.rst:48]: That works too!

**Comment by ankingcodes** [comment]
@ben-albrecht changes done. I realised I haven't added a test yet for this feature. Is it required ?

**Comment by ben-albrecht** [comment]
> I realised I haven't added a test yet for this feature. Is it required ?

Ah good catch. This could be tricky to test given the interactive nature. You might be able to get this working using a `.in` file (for stdin). Check out `test/library/standard/FileSystem/copyFile` as an example test using this feature.

**Comment by ankingcodes** [comment]
@ben-albrecht added the test. I think this PR is good for merge.

**Comment by Spartee** [comment]
I just did another cursory review without running the tests and I like that it repeats now. This is exactly as I would expect the command to act as a developer. LGTM 👍

File: test/mason/masonInteractive.stdin [hunk 15]
```diff
@@ -0,0 +1,4 @@
+project
```
Comment (ben-albrecht) [inline test/mason/masonInteractive.stdin:1]: Ah, _stdin_, not  _in_. Sorry to lead you down the wrong path! Glad you were able to figure it out though :)

Comment (ankingcodes) [reply test/mason/masonInteractive.stdin:1]: No problem. 👍

**Comment by ankingcodes** [comment]
@ben-albrecht 
> Version number of 1.2.3.4 was allowed by interactive session. I think we want
to give an error and let user retry.

I did give a check for this by reusing the `checkVersion` function at `MasonModify.chpl (line 296)`. Is that buggy ? 

> Should any other flags interact with mason interactive? E.g. --show, --no-vcs

I can't think of a reason why other flags should not be allowed with `--interactive`.  

> Trying to end the program with ^D instead of ^C results in an infinite loop of
output. It would be nice if we could handle ^D more gracefully (I think ^D should
terminate program as well if possible). I am not yet sure how to do this.

We do have an initial message to use '^C' to quit the session. 
Anyways, I found a hack to get rid of the loop. `readline` always ends with a `\n` but on pressing ^D,
EOF is encountered, due to which there's no `\n`. So, I added a condition to exit there.

> Design question: Should mason init defaults inherit from any existing
Mason.toml values?

That's a very nice idea. Adding this right now.

**Comment by ankingcodes** [comment]
> Design question: Should mason init defaults inherit from any existing
Mason.toml values?

@ben-albrecht  Thinking of this again, one of the ideas of `mason init` is also to not overwrite existing files. Also, since we are re-using `mason init` functions for interactive, they skip the Mason.toml file since it is already created. 
Although the idea you gave is very nice, thinking from an opposite perspective, the idea of `--interactive` is to fundamentally create a fresh `Mason.toml` file with inputs given by an user during the session. 
I had started working on this but it gets more challenging than it seems, and would require re-implementing many functions of `mason init` to allow overwriting. Also, there's another scenario, if an user gives a different input for a field which is already present in toml file along with an input for a field not present in toml file, `mason init` would skip the feild already present, and would add the field that's not present. 
I think for now, we could simply give an error message if `Mason.toml` is found and could create a feature request for this which could be done in the future. Let me know what you think about this.

**Comment by ben-albrecht** [comment]
> I think for now, we could simply give an error message if Mason.toml is found and could create a feature request for this which could be done in the future. Let me know what you think about this.

This makes sense to me. I think it is worth documenting this future work in a short issue, but the added value relative to the effort required is small enough that it can be put off for now.

**Comment by ankingcodes** [comment]
Cool ! I'll make a separate issue for this.

**Comment by ankingcodes** [comment]
@ben-albrecht Made all the requested changes 👍

File: tools/mason/MasonNew.chpl [hunk 16]
```diff
@@ -91,13 +112,153 @@ proc masonNew(args) throws {
...
+  var gotCorrectChapelVersion = false;
+  while(1){
+    try {
+      if !gotCorrectPackageName {
+        write("Package name ");
+        if isInit then write("(" + defPackageName + ") ");
+        write(": ");
+        IO.stdout.flush();
+        IO.stdin.readline(packageName);
+        if packageName == '' {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:150]: Does it make sense to wrap this check in a function, e.g. `exitOnEOF(packageName);` or `checkEOF(packageName);`

File: tools/mason/MasonHelp.chpl [hunk 17]
```diff
@@ -105,6 +105,7 @@ proc masonBuildHelp() {
 proc masonNewHelp() {
   writeln('Usage:');
   writeln('    mason new [options] <project name>');
+  writeln('    mason new --interactive');
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl:108]: Depending on our answer to https://github.com/chapel-lang/chapel/pull/15748#pullrequestreview-424805913 - we may want to update the help output.

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: @ben-albrecht I think we should give an error. ( just to keep it clean, and since package name is the first question anyway so no point in skipping it, we'd still require inputs for version and chplVersion.  )

File: tools/mason/MasonNew.chpl (hunk 16 above)
Comment (ankingcodes) [reply tools/mason/MasonNew.chpl:150]: Okay, I think it'd be much cleaner.

File: tools/mason/MasonHelp.chpl (hunk 17 above)
Comment (ben-albrecht) [reply tools/mason/MasonHelp.chpl:108]: On more thought, I am actually leaning towards supporting it and skipping the package step as a result (though still doing the error checking).

This will reduce the chance of the user stumbling over an error, and will also allow us to revert the documentation back into the pre-forked version that your originally had:

```
mason new [options] <project name>');
```

where `--interactive` is just another argument.

I see the merit in your argument of keeping the implementation cleaner. I think the resulting cleaner user interface will be worth it though.

File: tools/mason/MasonNew.chpl (hunk 16 above)
Comment (ben-albrecht) [reply tools/mason/MasonNew.chpl:150]: BTW - nice trick you found there. Works like a charm!

File: tools/mason/MasonNew.chpl [hunk 18]
```diff
@@ -231,6 +219,13 @@ considered if no input is given.""");
   return (packageName, version, chapelVersion);
 }
 
+proc exitOnEOF(parameter) {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:222]: A bonus of refactoring this out into a separate function is that you can add a doc-string and explain why we need this, since it's probably not clear from the code alone :)

Comment (ankingcodes) [reply tools/mason/MasonNew.chpl:222]: Okay, I'll make that change along with [that](https://github.com/chapel-lang/chapel/pull/15748#discussion_r435540692)

File: tools/mason/MasonHelp.chpl (hunk 17 above)
Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: Should I proceed with the error message here ?

Comment (ben-albrecht) [reply tools/mason/MasonHelp.chpl:108]: > On more thought, I am actually leaning towards supporting it and skipping the package step as a result (though still doing the error checking and throwing an error if the package name has any issues like we normally do for non-interactive mason new).

Thanks for clarifying. I was advocating for (c) from https://github.com/chapel-lang/chapel/pull/15748#pullrequestreview-424805913

Do you have any opposition to that approach?

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: Okay, should we do the same with `mason init` ?

Comment (ben-albrecht) [reply tools/mason/MasonHelp.chpl:108]: I think so, since the directory name == package name by default.

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: > Do you have any opposition to that approach?

I don't really oppose it but I feel if an user is using interactive, they must be new to Chapel development. So I don't really think they'd actually go for using a name within the command.
That was why I suggested (a). 
If we consider adding name within the command, there might still be a case when the user has actually added a name there by mistake, and they might want to change it. Considering that, I'd also say (b) is a good option. What do you think ?

Comment (ben-albrecht) [reply tools/mason/MasonHelp.chpl:108]: That's a good argument. I agree and think (b) is the most ideal.

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: I also had a thought while developing this feature about dropping the `--interactive` flag. I think it would be a good idea to start interactive session if an user simply types `mason new`(outside directory) or `mason init` (inside directory). As an user, I think its more effort to type the `interactive` option.

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: I think using this method will separate this feature completely from the previous versions of mason init & mason new. If an user uses a name or option, commands will work as they should.

Comment (ben-albrecht) [reply tools/mason/MasonHelp.chpl:108]: Oh, I really like that idea.

Just to be clear, `mason new` would not be impacted, but `mason init` behavior changes, right? Previously `mason init` just used the default values, and now it would start an interactive session. Not a big change, but a change nevertheless.

Curious what @Spartee and/or @krishnadey30 thinks.

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: >  Previously mason init just used the default values, and now it would start an interactive session.

For `mason init` we could simply check if the directory where it's called is empty or not. If its empty, we start an interactive session else it behaves normally.

Comment (Spartee) [reply tools/mason/MasonHelp.chpl:108]: I support this. This thought actually crossed my mind when I was playing around with it. 

@ben-albrecht Why should ``mason new`` not also default to interactive?

Comment (krishnadey30) [reply tools/mason/MasonHelp.chpl:108]: I am also in favor of this.

Comment (ben-albrecht) [reply tools/mason/MasonHelp.chpl:108]: @ankingcodes  - Sounds like you may proceed on this feature for `mason new`. Great idea!

For `mason init`, I am still uncertain on some details. It seems odd/unpredictable for the command to have totally different behaviors based on what files are present rather than arguments, e.g.

```sh
> mkdir Foo
> mason init
# interactive session
^C
> touch Mason.toml
> mason init
# automatically chooses defaults in non-interactive session
```

A few ways we could make this behavior more predictable are:

1. `mason init` is always interactive no matter what.
2. `mason init` supports the `--interactive` flag and is non-interactive by default (how the PR currently is)

Do you have any thoughts?

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: @ben-albrecht I'd go with (1) since the idea is to get rid of the flag. 
> mason init is always interactive no matter what.

A small problem with using interactive if Mason.toml is present is that, we will require it then to overwrite it with the values given by an user. I know this has been discussed and is potential future feature, but I also think it might be irritating for an user to pass through the entire interactive session just to initialise their project. So, that was why I think the method I suggested above is a better way to tackle this. 
I'd like to improve it as - if a `Mason.toml && src/module.chpl` file is present, simply initialise a project, otherwise, run interactive. The reason to consider those two files is because, they're the only ones affected by this feature. 

Another option that comes to mind is introducing a `--no-interactive` flag to override the interactive `mason init`. But I'm not a big fan of this option. 

**The clash between interactive and init is that, init doesnt support overwriting, while interactive supports overwriting.** Would you be in support of keeping this feature only for `mason new` ? If `mason new` is used outside a directory, it'd work normally. If `mason new` is used inside an empty directory, interactive works as it does for `mason init`. If it finds `Mason.toml` or `src/module.chpl` it prints an error informing users that mason new should be used only in an empty directory. 

I'd be interested to know your opinions on these options.

Comment (ben-albrecht) [reply tools/mason/MasonHelp.chpl:108]: > A small problem with using interactive if Mason.toml is present is that, we will require it then to overwrite it with the values given by an user. 

I think it makes sense for `mason init` to not overwrite any existing files/`Mason.toml` entries for the non-interactive mode where we are using the defaults with no user input. However, for interactive mode, it seems reasonable to overwrite the existing `Mason.toml` values, because the user is able to give input on what those overwritten values should store.

> I know this has been discussed and is potential future feature, but I also think it might be irritating for an user to pass through the entire interactive session just to initialise their project.

I think coming back to the earlier idea of using any pre-existing `Mason.toml` fields as a default value for the interactive session would alleviate this issue. That said, I understand this is a big change which we may not want to tackle in this PR.

> Another option that comes to mind is introducing a --no-interactive flag to override the interactive mason init. But I'm not a big fan of this option.

I feel similarly about this.

> Would you be in support of keeping this feature only for mason new ?

I think this would be a reasonable path forward for keeping the scope of this PR manageable. This would give us more time to think about how we want `mason init` to behave too.

So to recap, the proposal is:

- `mason new` with no _positional_ arguments (flags are OK) uses the new interactive mode
- `mason init` remains unchanged, but we will plan to support interactive mode in the future once we reach design consensus in a separate issue.

Does that sound good to you?

Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: Sounds good ! I'll go ahead and complete this PR now. 👍

**Comment by ankingcodes** [comment]
@ben-albrecht Made the changes discussed above. Please test it and let me know for any changes.

File: tools/mason/MasonHelp.chpl (hunk 17 above)
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl:108]: I think this should now be:

```suggestion
  writeln('    mason new [options]);
  ```
```

File: doc/rst/tools/mason/mason.rst [hunk 19]
```diff
@@ -35,7 +35,10 @@ Basic Usage
 Starting a New Package
 ~~~~~~~~~~~~~~~~~~~~~~
 
-To initialize a new mason package, run the ``mason new [ package name ] [ options ]`` command, for example::
+To initialize a new mason package, run ``mason new``. This starts an interactive session which walks an user
+through the process of creating a project using Mason. This is highly recommended for new users.
+
+A more advanced user may use the ``mason new [ package name ] [ options ]`` command, for example::
```
Comment (ben-albrecht) [inline doc/rst/tools/mason/mason.rst:41]: Might as well improve this and use the same syntax as the help output:

```suggestion
A more advanced user may use the ``mason new [options] <project name>`` command, for example::
```

File: tools/mason/MasonHelp.chpl [hunk 20]
```diff
@@ -118,6 +119,7 @@ proc masonInitHelp(){
   writeln('Usage:');
   writeln('    mason init [options] <directory name>');
   writeln('    mason init [options]');
+  writeln('    mason init --interactive');
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl:122]: This can be removed now

**Comment by ankingcodes** [comment]
> I think our support for args has been lost in the recent change as well as our nice clean --name error message:

@ben-albrecht  I believe we want to separate the interactive feature from the current working of mason new? 
Therefore, since this feature should simply be an addition to the current `mason new`, I'm not in support of using options. 
Also, Thinking from the perspective of a new user, since, they would be the ones using this feature, I don't really think they would want to use options with interactive.  
For `--name`, we already have an error message show up if an user inputs illegal name within the session. However I can improve the error message displayed on typing `mason new --name hello`

File: tools/mason/MasonHelp.chpl (hunk 17 above)
Comment (ankingcodes) [reply tools/mason/MasonHelp.chpl:108]: I think we should simply use 
```
writeln('   mason new        starts an interactive session');
or 
writeln('    mason new');
```

**Comment by ankingcodes** [comment]
Okay. We want to keep this feature very simple. We just want to start an interactive session when `mason new` is typed, instead of the help message that is current displayed. Nothing more than that. 
for example, lets consider the flag `--name`.
In current version of `mason new`, this flag is used as `mason new illegal-name --name legalName`.
So, we keep it like that, after our PR is merged. 
We dont want something like `mason new --name legalName` as it will be an error to the prev version.
Interactive should not affect the previous working of `mason new`.  Now there's a concern that an user might use illegal name in interactive, so for that we have an error that shows up within the session. 
I hope you're able to understand better now.

**Comment by ben-albrecht** [comment]
@ankingcodes - Understood. I'm on board with that. Thanks for explaining.

**Comment by ankingcodes** [comment]
@ben-albrecht Okay, made the final changes.

File: tools/mason/MasonNew.chpl [hunk 21]
```diff
@@ -152,24 +289,39 @@ proc addGitIgnore(dirName: string) {
...
-                     '[dependencies]' +
-                     '\n';
+proc getBaseTomlString(packageName: string, version: string, chapelVersion: string) {
+  const baseToml = """[brick]
+name = "%s"
+version = "%s"
+chplVersion = "%s"
+
+[dependencies]
+
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl:299]: This adds an extra newline more than before, which shows up in the test `test/mason/chplVersion/mason-chpl-version-new.chpl`.

We either need to update the `.good` file or update this to remove one of the newlines.

**Comment by ankingcodes** [comment]
@ben-albrecht While reviewing this PR, I found that at the last query `Is this okay? (Y/N)` , pressing `Enter` shows the TomlFile preview and the query again. Should that simply go forward with Y and create the project ?

**Comment by ben-albrecht** [comment]
> While reviewing this PR, I found that at the last query Is this okay? (Y/N) , pressing Enter shows the TomlFile preview and the query again. Should that simply go forward with Y and create the project ?

I like the `y/n` query for the final review step, since it helps prevent a user accidentally skipping over it. I think the current behavior is good.

**Comment by ankingcodes** [comment]
@ben-albrecht Okay, this is good for merge then

**Comment by ben-albrecht** [comment]
> Okay, this is good for merge then

Agreed!  As a heads up, there were a few mason-related test failures over the weekend (due to system changes). I'd like to make sure they come in clean today before merging this.

**Comment by ben-albrecht** [comment]
FYI - still waiting for the mason failures to clear up...

**Comment by ben-albrecht** [comment]
Looks like the tests have cleared up. Merging
